hedera-node/src/main/java/com/hedera/services/txns/ethereum/EthereumTransitionLogic.java

```diff
@@ -0,0 +1,275 @@
+			if (ethTxData.nonce() != accountNonce) {
+				return ResponseCodeEnum.WRONG_NONCE;
+			}
+		}
+
+		if (txBody.hasContractCall()) {
+			return contractCallTransitionLogic.semanticCheck().apply(txBody);
+		} else if (txBody.hasContractCreateInstance()) {
+			return contractCreateTransitionLogic.semanticCheck().apply(txBody);
+		} else {
+			// This should only happen when we update the createSyntheticTransactionBody
+			// and then forget to update this code.
+			return ResponseCodeEnum.FAIL_INVALID;
+		}
+	}
+
+	@Override
+	public Function<TransactionBody, ResponseCodeEnum> semanticCheck() {
+		// since we override `validateSemantics` this is to hedge against this accidentally being called
+		// and ensure we always fail if we depend on this overridden path.
+		return ignore -> ResponseCodeEnum.NOT_SUPPORTED;
+	}
+
+	@Override
+	public void preFetch(final TxnAccessor accessor) {
+		var ethTxData = spanMapAccessor.getEthTxDataMeta(accessor);
+
+		EthereumTransactionBody op = accessor.getTxn().getEthereumTransaction();
+		ethTxData = maybeUpdateCallData(accessor, ethTxData, op);
+
+		var ethTxSigs = EthTxSigs.extractSignatures(ethTxData);
+		spanMapAccessor.setEthTxSigsMeta(accessor, ethTxSigs);
+
+		TransactionBody txBody = createSyntheticTransactionBody(ethTxData);
+
+		spanMapAccessor.setEthTxBodyMeta(accessor, txBody);
+		if (txBody.hasContractCall()) {
+			contractCallTransitionLogic.preFetchOperation(txBody.getContractCall());
+		}
+	}
+
+	private EthTxData maybeUpdateCallData(final TxnAccessor accessor, EthTxData ethTxData,
+			final EthereumTransactionBody op) {
+		if ((ethTxData.callData() == null || ethTxData.callData().length == 0) && op.hasCallData()) {
+			var callDataFileId = op.getCallData();
+			validateTrue(hfs.exists(callDataFileId), INVALID_FILE_ID);
+			byte[] callDataFile = Hex.decode(hfs.cat(callDataFileId));
+			validateFalse(callDataFile.length == 0, CONTRACT_FILE_EMPTY);
+			ethTxData = ethTxData.replaceCallData(callDataFile);
+			spanMapAccessor.setEthTxDataMeta(accessor, ethTxData);
+		}
+		return ethTxData;
+	}
+
+	private TransactionBody createSyntheticTransactionBody(EthTxData ethTxData) {
+		// maybe we could short circuit direct calls to tokens and topics?  maybe not
+		if (ethTxData.to() != null && ethTxData.to().length != 0) {
+			var synthOp = ContractCallTransactionBody.newBuilder()
+					.setFunctionParameters(ByteString.copyFrom(ethTxData.callData()))
+					.setGas(ethTxData.gasLimit())
+					.setAmount(ethTxData.value().divide(WEIBARS_TO_TINYBARS).longValueExact())
+					.setContractID(EntityIdUtils.contractIdFromEvmAddress(ethTxData.to())).build();
+			return TransactionBody.newBuilder().setContractCall(synthOp).build();
+		} else {
+			final var autoRenewPeriod =
+					Duration.newBuilder().setSeconds(globalDynamicProperties.minAutoRenewDuration()).build();
+			var synthOp = ContractCreateTransactionBody.newBuilder()
+					.setInitcode(ByteString.copyFrom(ethTxData.callData()))
```

Comment:
Q: Is this prehydration? If so then it'd always be empty when the initCode is in a file.
Should there be a conditional to populate fileID on empty callData?

Reply:
* This synthetic operation is not what is persisted in the record file.  It's only used to drive the transition logic.
* The ethTxData is the rehydrated version, if it had a fileId it was inserted
* We require a rehydrated EthTxData for signature validation to work.

Reply:
Gotcha.